Subject: [PATCH] Fix broken RST handling in ip_conntrack
Date:	Tue, 04 Jan 2005 04:19:30 +0000
X-BK-ChangeSetKey: gandalf@netfilter.org[torvalds]|ChangeSet|20050104041930|38497
Patch-mainline: yes
From: olh@suse.de


ChangeSet 1.2136.3.34, 2005/01/03 20:19:30-08:00, gandalf@netfilter.org

	[PATCH] Fix broken RST handling in ip_conntrack
	
	Here's a patch that fixes a pretty serious bug introduced by a recent
	"bugfix".  The problem is that RST packets are ignored if they follow an
	ACK packet, this means that the timeout of the connection isn't decreased,
	so we get lots of old connections lingering around until the timeout
	expires, the default timeout for state ESTABLISHED is 5 days.
	
	This needs to go into -bk as soon as possible.  The bug is present in
	2.6.10 as well.
	
	Signed-off-by: Andrew Morton <akpm@osdl.org>
	Signed-off-by: Linus Torvalds <torvalds@osdl.org>



 ip_conntrack_proto_tcp.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletion(-)


diff -Nru a/net/ipv4/netfilter/ip_conntrack_proto_tcp.c b/net/ipv4/netfilter/ip_conntrack_proto_tcp.c
--- a/net/ipv4/netfilter/ip_conntrack_proto_tcp.c	2005-01-03 22:14:55 -08:00
+++ b/net/ipv4/netfilter/ip_conntrack_proto_tcp.c	2005-01-03 22:14:55 -08:00
@@ -906,7 +906,8 @@
 		if (index == TCP_RST_SET
 		    && ((test_bit(IPS_SEEN_REPLY_BIT, &conntrack->status)
 		         && conntrack->proto.tcp.last_index <= TCP_SYNACK_SET)
-		        || conntrack->proto.tcp.last_index == TCP_ACK_SET)
+		        || (!test_bit(IPS_ASSURED_BIT, &conntrack->status)
+			 && conntrack->proto.tcp.last_index == TCP_ACK_SET))
 		    && after(ntohl(th->ack_seq),
 		    	     conntrack->proto.tcp.last_seq)) {
 			/* Ignore RST closing down invalid SYN or ACK
-
To unsubscribe from this list: send the line "unsubscribe bk-commits-head" in
the body of a message to majordomo@vger.kernel.org
More majordomo info at  http://vger.kernel.org/majordomo-info.html

